### PR TITLE
TandemFoil: shorter cosine cycles T_max=5 at lr=2e-4

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The current TandemFoil baseline uses T_max=10 with lr=2e-4. On AirfRANS, T_max=5 (shorter cosine cycles) was part of the golden config that achieved -22.3% improvement. More frequent LR restarts create more basin-hopping opportunities, potentially helping the Lion optimizer escape shallow minima. We expect T_max=5 to improve surface pressure MAE on TandemFoil.

## Instructions
Run the baseline TandemFoil command with T_max shortened to 5:

```bash
SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset tandemfoil \
  --optimizer lion \
  --lr 2e-4 \
  --cosine-t-max 5 \
  --no-use-ema \
  --model-slices 64 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --wandb_group tandem-lr2e4-tmax \
  --epochs 9999
```

Changes vs baseline:
- `--cosine-t-max 5` (was 10)

## Baseline
Current best TandemFoil metrics (PR #2610, W&B: ixs1rqgk):
- val_primary/surface_pressure_mae: **45.07** (lower is better)

Reproduce baseline:
```bash
cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 2e-4 --cosine-t-max 10 --no-use-ema --model-slices 64 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT sakura:

## Results

### Key Metrics (Best Epoch 49 of 100+)

| Metric | Baseline (PR #2610) | This Run | Delta |
|--------|---------------------|----------|-------|
| **val_primary/surface_pressure_mae** | **45.07** | **53.94** | **+19.7% (worse)** |
| val_eq4/surface_pressure_mae | 45.07 | 53.94 | +19.7% |
| val_geom_camber_cruise/surface_pressure_mae | — | 37.79 | — |
| val_geom_camber_rc/surface_pressure_mae | — | 65.59 | — |
| val_re_rand/surface_pressure_mae | — | 51.22 | — |
| val_single_in_dist/surface_pressure_mae | — | 61.15 | — |

### Experiment Details

- **Command:**
```bash
SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset tandemfoil --optimizer lion --lr 2e-4 --cosine-t-max 5 \
  --no-use-ema --model-slices 64 --enable-fourier --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition \
  --wandb-group tandem-lr2e4-tmax --wandb-name "sakura/tandem-lr2e4-tmax5" \
  --agent sakura --epochs 9999
```
- **W&B Run ID:** `37yu0ure` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/37yu0ure))
- **Peak VRAM:** ~41.9 GB
- **Total Epochs:** 100+ (productive training: epochs 1-49, gradient explosion: epochs 53-86, frozen at Infinity: epochs 86+)

### What Happened

**T_max=5 delivered strong convergence for ~50 epochs but ultimately suffered the same gradient explosion as T_max=30, just delayed.**

**Convergence phase (epochs 1-49):**
- Beautiful staircase pattern: each 5-epoch cosine cycle pushed the val floor lower
- Cycle-end troughs: 110→88→72→60→56→54 (each cycle ~5-10% improvement)
- Best val_primary = 53.94 at epoch 49 (vs baseline 45.07, +19.7% worse)
- Grad norm remained stable at 0.9-2.0 through most of this phase

**Instability onset (epochs 51-55):**
- Grad norm spiked: 1.57 → 3.39 → 7.81 at epochs 52-54
- Short T_max=5 cycle initially contained it — grad norm dropped back to 3.49 after cycle reset

**Escalating instability (epochs 55-72):**
- Each cosine cycle's peak grad norm was higher than the last: 7.81 → 10.72 → 23.53 → ...
- The short cycles provided some containment but couldn't prevent the underlying exponential growth

**Full explosion (epochs 72-86):**
- Grad norm: 288 → 531 → 1,032 → 1,913 → 3,780 → ... → 1 trillion → Infinity
- Val degraded from ~55 to ~115
- Epoch 86: grad_norm = Infinity, model frozen

**Frozen phase (epochs 86+):**
- All metrics identical, optimizer producing no weight updates
- Will run until 180-min timeout (~epoch 150)

### Analysis

1. **T_max=5 is much more stable than T_max=30** — it provided ~50 productive epochs vs T_max=30's ~40 before instability, and the short cycles successfully contained early gradient spikes that would have been catastrophic with T_max=30.

2. **But T_max=5 cannot fully prevent gradient explosions at lr=2e-4.** The instability is a property of the learning rate and model architecture, not just the cycle length. The short cycles delay but don't eliminate the issue.

3. **Best result (53.94) is +19.7% above baseline (45.07).** This is a single run starting from scratch — the baseline was achieved with the full tuned setup. The convergence trajectory was excellent and may have continued improving if not for the explosion.

4. **The experiment strongly validates the need for gradient clipping.** The `--grad-clip` flag is now available in train.py. Using `--grad-clip 1.0` would likely have allowed productive training to continue well beyond epoch 49.

### Suggested Follow-ups

1. **T_max=5 + grad-clip 1.0**: Re-run this exact configuration with `--grad-clip 1.0` to prevent the explosion and allow continued convergence past epoch 49. This is the most obvious win.
2. **T_max=5 + lower lr (1e-4)**: The instability is ultimately driven by the learning rate. Lower lr with short cycles could avoid the explosion entirely, though convergence would be slower.
3. **T_max=3**: Even shorter cycles for more frequent resets. Could provide tighter containment, though each cycle gets less time to descend.
4. **Cosine warmup restart with decay**: Reduce the peak LR after each restart (e.g., 0.9x multiplier). This would naturally control the instability as training progresses.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2773 because ADVISOR: Closing — T_max=5 at lr=2e-4 reached 53.94 vs updated baseline 44.72 (20.6% worse). Training diverged at epoch 49 with gradient explosion (grad norms reached infinity by ep86). T_max=5 delays but cannot prevent gradient instability at lr=2e-4. T_max=10 remains optimal for TandemFoil. Excellent analysis of the divergence dynamics — confirms gc=1.0 would be needed with shorter cycles..
